mono/metadata/class.c

```diff
@@ -6541,6 +6541,8 @@ mono_bounded_array_class_get (MonoClass *eclass, guint32 rank, gboolean bounded)
 	GSList *list, *rootlist = NULL;
 	int nsize;
 	char *name;
+	/* use image set for element classes of type generic inst and array, as they may contain generic arguments from other images */
+	MonoImageSet* image_set = (eclass->class_kind == MONO_CLASS_GINST || eclass->class_kind == MONO_CLASS_ARRAY) ? mono_metadata_get_image_set_for_class (eclass) : NULL;
```

Comment:
This will put `int[][]` into an image set.  Is that a problem?

Reply:
It's not intuitive, but it doesn't seem to cause problems. It's an image set of size 1 (just mscorlib). This is the same image set as would be used for `List<int>` AFAIK.

I could avoid this, but it seems to require an expensive check in `mono_bounded_array_class_get` or storing result of that check in a bit on each `MonoClass` as to whether it references types from other images.

Reply:
Not really asking for it to be done in this PR, but in general: if image sets of size 1 are fine, why don't we make all mempools owned by image sets, not images and simplify all this logic everywhere we can  (although I guess computing image sets is pricy)